utils/cmake/dreamcast.toolchain.cmake

```diff
@@ -0,0 +1,106 @@
+# CMake Toolchain file for targeting the Dreamcast or NAOMI with CMake.
+#  (c)2023 Falco Girgis, Luke Benstead
+#
+# This file is to be passed to CMake when compiling a regular CMake project
+# to cross-compile for the Dreamcast, using the KOS environment and settings.
+#
+#     cmake /path/to/src -DCMAKE_TOOLCHAIN_FILE=${KOS_BASE}/kos/utils/cmake/dreamcast.toolchain.cmake
+#
+#   or alternatively:
+#
+#     cmake /path/to/src -DCMAKE_TOOLCHAIN_FILE=${KOS_CMAKE_TOOLCHAIN}
+#
+# The toolchain enables the options required for the compiler to use the 
+# special SH4 instructions for sin/cos and invSqrt. Note that even when 
+# these options are present, they are not activated unless -ffast-math 
+# is also provied. Typically you will want to provide a project-level 
+# option for enabling this flag, which will make the toolchain's SH4 
+# instruction flags also take effect.
+#
+# Frame pointers are enabled in debug builds as these are required for 
+# stack traces and GDB. They are disabled in release.
+# 
+# The original toolchain file was created by Kazade for the Simulant 
+# engine who has graciously allowed the rest of the scene to warez it. 
+#
+# Special thanks to Matt Godbolt of Compiler Explorer, for supporting
+# the SH toolchains, which allowed us to figure out the precise magical
+# incantations of compiler flags for fast math.  
+
+##### Configure CMake System #####
+set(CMAKE_SYSTEM_NAME Generic)
+set(CMAKE_SYSTEM_VERSION 1)
+set(PLATFORM_DREAMCAST TRUE)
+
+##### Configure Cross-Compiler #####
+set(CMAKE_CROSSCOMPILING TRUE)
+set(CMAKE_C_COMPILER $ENV{KOS_CC_BASE}/bin/sh-elf-gcc)
+set(CMAKE_CXX_COMPILER $ENV{KOS_CC_BASE}/bin/sh-elf-g++)
+
+set(CMAKE_EXECUTABLE_SUFFIX ".elf")
```

Comment:
I've been testing with this toolchain file and have come across that this doesn't apply properly. According to this (https://discourse.cmake.org/t/how-to-specify-executable-file-type-for-arm/4896/3) setting "CMAKE_EXECUTABLE_SUFFIX" get overwritten when set in the toolchain file. Seems like our options are:

1) Set "CMAKE_EXECUTABLE_SUFFIX_C", "CMAKE_EXECUTABLE_SUFFIX_CXX", and "CMAKE_EXECUTABLE_SUFFIX_ASM" variables to ".elf" which do not overwritten and does fix this, but seems to be an unrecommended (but common for bare-metal) configuration.

2) The other option is to set "CMAKE_SYSTEM_NAME" to "Generic-ELF" from "Generic" which is just Generic but with the elf suffix appended to executables. The the downside is that it was merged in for CMake 3.23 (https://gitlab.kitware.com/cmake/cmake/-/merge_requests/6630) which has only been out about a year.

I've tested both of these spit out executables with the correct ".elf" extension, just a question of if we want to do it the new correct way, or the older hack which will support more versions of CMake.

Reply:
WELL NO WONDER this behavior seemed to be either not working or not working consistently for me. Thank you! 

You bring up a good point about which CMake version minimum we're targeting here...